File: 2022/Python/day03/day03.py
import functools


with open('day03.txt', 'r') as f :
    rucksacks = f.readlines()


def itemPriority(item) :
    if item.isupper() :
        return ord(item) - ord('A') + 27
    else :
        return ord(item) - ord('a') + 1


def solvePartOne() :
    prioSum = 0
    for rucksack in rucksacks :
        size = len(rucksack) // 2
        comp1, comp2 = set(rucksack[:size]), set(rucksack[size:])
        commonItem = comp1.intersection(comp2).pop()
        prioSum += itemPriority(commonItem)

    print(f"Part 1. Sum = {prioSum}")


def solvePartTwo() :
    prioSum = 0

    groups = [ rucksacks[i:i+3] for i in range(0, len(rucksacks), 3) ]
    for group in groups :
        sackSets = [set(sack.strip()) for sack in group]
        commonItem = functools.reduce(lambda acc, sack: acc.intersection(sack), sackSets).pop()
        prioSum += itemPriority(commonItem)

    print(f"Part 2. Sum = {prioSum}")


solvePartOne()
solvePartTwo()

File: 2022/Kotlin/src/main/kotlin/day03/Day03.kt
package day03

import java.io.File

val rucksacks: List<String> = File("src/main/resources/day03.txt").readLines()

fun Char.itemPriority(): Int = if (this.isUpperCase()) (this - 'A' + 27) else (this - 'a' + 1)

fun solvePartOne() {
    val prioritySum: Int = rucksacks.sumOf { rucksack ->
        val compSize = rucksack.length / 2
        val firstCompartment = rucksack.take(compSize).toSet()
        val secondCompartment = rucksack.takeLast(compSize).toSet()
        val sharedItem: Char = firstCompartment.intersect(secondCompartment).first()

        sharedItem.itemPriority()
    }
    println("Part 1. Sum = $prioritySum")
}

fun solvePartTwo() {
    val prioritySum: Int = rucksacks.chunked(3).sumOf { elfGroup ->
        val rucksackSets = elfGroup.map { rucksack -> rucksack.toSet() }
        val sharedItem: Char = rucksackSets.reduce { acc, sack -> acc.intersect(sack) }.first()

        sharedItem.itemPriority()
    }

    println("Part 2. Sum = $prioritySum")
}

fun main() {
    solvePartOne()
    solvePartTwo()
}